pkg/test/framework.go

```diff
@@ -40,36 +40,45 @@ import (
 )
 
 var (
-	// Global framework struct
+	// Global is a global framework struct that the test can use.
 	Global *Framework
 	// mutex for AddToFrameworkScheme
 	mutex = sync.Mutex{}
-	// whether to run tests in a single namespace
-	singleNamespace *bool
+	// singleNamespaceInternal determines whether tests are to be run in a single namespace or not.
+	singleNamespaceInternal bool
 	// decoder used by createFromYaml
 	dynamicDecoder runtime.Decoder
 	// restMapper for the dynamic client
 	restMapper *restmapper.DeferredDiscoveryRESTMapper
 )
 
+// Framework contains all relevant variables needed for running tests with the operator-sdk.
 type Framework struct {
 	Client            *frameworkClient
 	KubeConfig        *rest.Config
 	KubeClient        kubernetes.Interface
 	Scheme            *runtime.Scheme
-	NamespacedManPath *string
+	NamespacedManPath string
 	Namespace         string
 	LocalOperator     bool
 }
 
-func setup(kubeconfigPath, namespacedManPath *string, localOperator bool) error {
-	namespace := ""
-	if *singleNamespace {
-		namespace = os.Getenv(TestNamespaceEnv)
+// Setup initializes the Global.Framework variable and its fields.
+func Setup(kubeconfigPath, namespacedManPath, namespace string, singleNamespace, localOperator bool) error {
```

Comment:
I'm not sure if there are problematic implications, but can this function signature be refactored to something like:

```go
type FrameworkOptions struct {
    KubeconfigPath string
    NamespacedManPath string
    Namespace string
    SingleNamespace bool
    LocalOperator bool
}

func NewFramework(opts FrameworkOptions) (*Framework, error)
```

Then `AddToFrameworkScheme()` could be a method on the `Framework` struct and renamed to `AddToScheme()`.

I realize this would be a breaking change, so that's worth some discussion. 

I think we should try to avoid having global mutable state as much as possible and favor returning and passing values around instead.

Reply:
The biggest difficulty with not having the global `Framework` variable is how the framework would be passed around in the tests. We don't have a simple way to pass down variables through tests without global variables and having the global variable makes it easier for us to implement the main entry that sets up a single framework that all tests run after the entry can use.

Reply:
One way to avoid the global state in `pkg/test` would be to have a global framework variable in the user's `main_test.go` file that gets set to the result of `NewFramework`. But this would be a pretty important breaking change that we would have to make very clear to users.

Reply:
Ok I think I'm understanding this a little bit better now having taken a look at https://github.com/operator-framework/operator-sdk/blob/master/test/test-framework/test/e2e/memcached_test.go.

If I understand correctly, right now it's basically required for tests to call `f.MainEntry(m)` to get `Global` setup properly. My concern is that if we open up more of the test framework and make it possible to run tests without calling `f.MainEntry`, then users will have to know the right order of calls to make before `Global` is safe to use.

If getting rid of the exported `Global` variable will cause too big of a breaking change, would it be possible to make its zero value (or initialized value) usable?